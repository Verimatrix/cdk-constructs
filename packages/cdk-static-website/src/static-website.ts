import {
  AliasConfiguration,
  CloudFrontWebDistribution,
  CloudFrontWebDistributionProps,
  LambdaFunctionAssociation,
  CfnDistribution,
} from '@aws-cdk/aws-cloudfront';
import { CloudFrontTarget } from '@aws-cdk/aws-route53-targets';
import { Construct } from '@aws-cdk/core';

import { WebsiteAliasRecord } from './website-alias-record';
import { WebsiteBucket, WebsiteBucketProps } from './website-bucket';

export interface AliasProps extends AliasConfiguration {
  /** The domain name for the site like 'example.com' */
  readonly domainName: string;

  /**
   * Allow searching a private hosted zone.
   * @default false
   */
  readonly privateZone?: boolean;

  /**
   * Specifies the ID of the VPC associated with a private hosted zone.
   *
   * If a VPC ID is provided and privateZone is false, no results will be returned
   * and an error will be raised
   *
   * @default - No VPC ID
   */
  readonly vpcId?: string;
}

export interface StaticWebsiteProps extends WebsiteBucketProps {
  /**
   * BucketConfiguration is used to configured the S3 website bucket
   *
   * @default - The website bucket is provided with default values
   */
  readonly bucketConfiguration?: WebsiteBucketProps;

  /**
   * AliasConfiguration is used to configured CloudFront to respond to requests on custom domain names.
   *
   * @default - No custom domain names are set up
   */
  readonly aliasConfiguration?: AliasProps;

  /**
   * AWS WAF WebACL to associate with this CloudFront distribution
   *
   * @default - No AWS Web Application Firewall web access control list (web ACL)
   */
  readonly webACLId?: string;

  /**
   * An override flag that allows you to turn off support for IPv6 if required.
   *
   * @default - Cloudfront IPv6 support is enabled and if you've supplied an aliasConfiguration, an
   * AAAA record will be created for your service, set this to true to switch this off.
   */
  readonly disableIPv6?: boolean;

  /**
   * How CloudFront should handle requests that are not successful (e.g., PageNotFound).
   *
   * @default - No custom error responses.
   */
  readonly errorConfigurations?: CfnDistribution.CustomErrorResponseProperty[];
}

export class StaticWebsite extends Construct {
  readonly distribution: CloudFrontWebDistribution;

  constructor(scope: Construct, id: string, props: StaticWebsiteProps = {}) {
    super(scope, id);

    const { aliasConfiguration, bucketConfiguration, webACLId, disableIPv6, errorConfigurations } = props;

    const websiteBucket = new WebsiteBucket(this, 'Bucket', {
      ...bucketConfiguration,
    });

    const distributionConfig: CloudFrontWebDistributionProps = {
      webACLId,
      enableIpV6: !disableIPv6,
      originConfigs: [
        {
          s3OriginSource: {
            ...websiteBucket.s3OriginConfig,
          },
          behaviors: [{ isDefaultBehavior: true }],
        },
      ],
      aliasConfiguration,
      errorConfigurations,
    };

    this.distribution = new CloudFrontWebDistribution(this, 'Distribution', distributionConfig);

    if (aliasConfiguration) {
      new WebsiteAliasRecord(this, 'AliasRecord', {
        domainName: aliasConfiguration.domainName,
        recordNames: aliasConfiguration.names,
        target: new CloudFrontTarget(this.distribution),
        disableIPv6,
        privateZone: aliasConfiguration.privateZone,
        vpcId: aliasConfiguration.vpcId,
      });
    }
  }

  public addLambdaFunctionAssociation(assosiation: LambdaFunctionAssociation): void {
    this.addLambdaFunctionAssociations([assosiation]);
  }

  public addLambdaFunctionAssociations(assosiations: LambdaFunctionAssociation[]): void {
    const cfDist = this.distribution.node.findChild('CFDistribution') as CfnDistribution;

    cfDist.addOverride(
      'Properties.DistributionConfig.DefaultCacheBehavior.LambdaFunctionAssociations',
      assosiations.map((assosiation: LambdaFunctionAssociation) => ({
        EventType: assosiation.eventType,
        LambdaFunctionARN: assosiation.lambdaFunction.functionArn,
      })),
    );
  }
}
