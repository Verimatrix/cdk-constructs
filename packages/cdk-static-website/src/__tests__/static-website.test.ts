import { Stack } from 'aws-cdk-lib';
import 'jest-cdk-snapshot';

import { StaticWebsite } from '../static-website';

test('default setup', (): void => {
  const stack = new Stack();

  new StaticWebsite(stack, 'StaticWebsite', {
    disableUpload: true,
  });

  expect(stack).toMatchCdkSnapshot({
    ignoreAssets: true,
  });
});

// test('lambda at edge', (): void => {
//   const stack = new Stack();

//   const staticWebsite = new StaticWebsite(stack, 'StaticWebsite', {
//     disableUpload: true,
//   });

//   staticWebsite.addLambdaFunctionAssociation({
//     eventType: aws_cloudfront.LambdaEdgeEventType.ORIGIN_REQUEST,
//     lambdaFunction: aws_lambda.Version.fromVersionArn(stack, 'LambdaEdge', 'arn:aws:lambda:us-east-1:123456789012:function:my-function:1'),
//   });

//   expect(stack).toMatchCdkSnapshot({
//     ignoreAssets: true,
//   });
// });

// test('test sertting errorConfigurations', (): void => {
//   const stack = new Stack();

//   new StaticWebsite(stack, 'StaticWebsite', {
//     disableUpload: true,
//     errorConfigurations: [
//       {
//         errorCode: 404,
//         errorCachingMinTtl: 3,
//         responseCode: 200,
//         responsePagePath: '/index.html',
//       },
//     ],
//   });

//   expect(stack).toMatchCdkSnapshot({
//     ignoreAssets: true,
//   });
// });
