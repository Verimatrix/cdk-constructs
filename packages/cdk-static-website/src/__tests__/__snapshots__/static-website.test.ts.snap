// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default setup 1`] = `
Object {
  "Parameters": Any<Object>,
  "Resources": Object {
    "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F": Object {
      "DependsOn": Array [
        "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
      ],
      "Properties": Object {
        "Code": Any<Object>,
        "Description": Object {
          "Fn::Join": Array [
            "",
            Array [
              "Lambda function for auto-deleting objects in ",
              Object {
                "Ref": "StaticWebsiteBucket0E92E0FC",
              },
              " S3 bucket.",
            ],
          ],
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "StaticWebsiteBucket0E92E0FC": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "aws-cdk:auto-delete-objects",
            "Value": "true",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "StaticWebsiteBucketAutoDeleteObjectsCustomResource840785FA": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "StaticWebsiteBucketPolicy86C0E5CF",
      ],
      "Properties": Object {
        "BucketName": Object {
          "Ref": "StaticWebsiteBucket0E92E0FC",
        },
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3AutoDeleteObjects",
      "UpdateReplacePolicy": "Delete",
    },
    "StaticWebsiteBucketPolicy86C0E5CF": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "StaticWebsiteBucket0E92E0FC",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::GetAtt": Array [
                    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
                    "Arn",
                  ],
                },
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "StaticWebsiteBucket0E92E0FC",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "StaticWebsiteBucket0E92E0FC",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": Object {
                  "Fn::GetAtt": Array [
                    "StaticWebsiteDistributionOrigin1S3Origin3045572F",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "StaticWebsiteBucket0E92E0FC",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "StaticWebsiteDistributionBAD21F75": Object {
      "Properties": Object {
        "DistributionConfig": Object {
          "DefaultCacheBehavior": Object {
            "AllowedMethods": Array [
              "GET",
              "HEAD",
            ],
            "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
            "Compress": true,
            "ResponseHeadersPolicyId": Object {
              "Ref": "StaticWebsiteResponseHeadersPolicyF3EBE566",
            },
            "TargetOriginId": "StaticWebsiteDistributionOrigin1128FCAFB",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": Array [
            Object {
              "DomainName": Object {
                "Fn::GetAtt": Array [
                  "StaticWebsiteBucket0E92E0FC",
                  "RegionalDomainName",
                ],
              },
              "Id": "StaticWebsiteDistributionOrigin1128FCAFB",
              "S3OriginConfig": Object {
                "OriginAccessIdentity": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "origin-access-identity/cloudfront/",
                      Object {
                        "Ref": "StaticWebsiteDistributionOrigin1S3Origin3045572F",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "PriceClass": "PriceClass_100",
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "StaticWebsiteDistributionOrigin1S3Origin3045572F": Object {
      "Properties": Object {
        "CloudFrontOriginAccessIdentityConfig": Object {
          "Comment": "Identity for StaticWebsiteDistributionOrigin1128FCAFB",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "StaticWebsiteResponseHeadersPolicyF3EBE566": Object {
      "Properties": Object {
        "ResponseHeadersPolicyConfig": Object {
          "CustomHeadersConfig": Object {
            "Items": Array [],
          },
          "Name": "StaticWebsiteResponseHeadersPolicy53E0361F",
          "SecurityHeadersConfig": Object {
            "ContentSecurityPolicy": Object {
              "ContentSecurityPolicy": "default-src 'none'; img-src 'self'; script-src 'self'; style-src 'self' 'unsafe-inline'; object-src 'none'; connect-src 'self'",
              "Override": true,
            },
            "ContentTypeOptions": Object {
              "Override": true,
            },
            "FrameOptions": Object {
              "FrameOption": "DENY",
              "Override": true,
            },
            "ReferrerPolicy": Object {
              "Override": true,
              "ReferrerPolicy": "same-origin",
            },
            "StrictTransportSecurity": Object {
              "AccessControlMaxAgeSec": 31536000,
              "IncludeSubdomains": true,
              "Override": true,
              "Preload": true,
            },
            "XSSProtection": Object {
              "ModeBlock": true,
              "Override": true,
              "Protection": true,
            },
          },
        },
      },
      "Type": "AWS::CloudFront::ResponseHeadersPolicy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
